tests/unit/challenges/test_utils.py

```diff
@@ -1,18 +1,182 @@
+        bad_response = utils.is_user_in_allowed_email_domains(bad_email, self.challenge.pk)
+        assert response
+        assert ~bad_response
+
+    def test_is_user_in_blocked_email_domains_when_true(self):
+        blocked_email = "useremail@badtest.com"
+        good_email = "useremail@test.com"
+        blocked_response = utils.is_user_in_blocked_email_domains(blocked_email, self.challenge.pk)
+        good_response = utils.is_user_in_blocked_email_domains(good_email, self.challenge.pk)
+        assert blocked_response
+        assert ~good_response
+
+
+@mock_ecr
+@mock_sts
+class TestWithAWSClients(BaseTestCase):
+    def setup(self):
+        super(TestWithAWSClients, self).setup()
+
+        self.logger = logging.getLogger(__name__)
+
+    @mock.patch("base.utils.get_boto3_client")
+    def test_get_or_create_ecr_repository_when_repository_exists(self, client):
+        client.return_value = self.ecr_client
+        expected = self.ecr_client.create_repository(repositoryName="TestRepo")
+        self.aws_keys["AWS_ACCOUNT_ID"] = expected["repository"]["registryId"]
+        response = utils.get_or_create_ecr_repository("TestRepo", self.aws_keys)
+        assert expected["repository"] == response[0]
+        self.aws_keys["AWS_ACCOUNT_ID"] = self.challenge.aws_account_id
+
+    @mock.patch("base.utils.get_boto3_client")
+    def test_get_or_create_ecr_repository_when_repository_does_not_exist(self, client):
+        client.return_value = self.ecr_client
+        response = utils.get_or_create_ecr_repository("TestRepo", self.aws_keys)
+        expected = self.ecr_client.describe_repositories(repositoryNames=["TestRepo"])
+        assert response == (expected["repositories"][0], True)
+    '''
+    @mock.patch("base.utils.get_boto3_client")
+    @mock.patch("logging.Logger.exception")
+    @mock.patch("botocore.client")
+    def test_get_or_create_ecr_repository_exceptions(self, client, mock_logger, get_client):
+        get_client.return_value = client
+        err_message = {"Error": {"Code": 406}}
+        e = ClientError(err_message, "test")
+        client.return_value = e
+        response = utils.get_or_create_ecr_repository("TestRepo", self.aws_keys)
+        print(response)
+        mock_logger.assert_called_with(e)
+    '''
+
+    @mock.patch("base.utils.get_boto3_client")
+    def test_create_federated_user(self, client):
+        client.return_value = self.sts_client
+        expected = {
+            'Credentials': {
+                'AccessKeyId': 'AKIAIOSFODNN7EXAMPLE',
+                'SecretAccessKey': 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYzEXAMPLEKEY',
+                'SessionToken': 'AQoDYXdzEPT//////////wEXAMPLEtc764bNrC9SAPBSM22wDOk4x4HIZ8j4FZTwdQWLWsKWHGBuFqwAeMicRXmxfpSPfIeoIYRqTflfKD8YUuwthAx7mSEI/qkPpKPi/kMcGdQrmGdeehM4IC1NtBmUpp2wUE8phUZampKsburEDy0KPkyQDYwT7WZ0wq5VSXDvp75YU9HFvlRd8Tx6q6fE8YQcHNVXAkiY9q6d+xo0rKwT38xVqr7ZD0u0iPPkUL64lIZbqBAz+scqKmlzm8FDrypNC9Yjc8fPOLn9FX9KSYvKTr4rvx3iSIlTJabIQwj2ICCR/oLxBA==',
```

Comment:
How about having these data generated randomly?

Reply:
Is there a benefit for having it random?

Reply:
Having Random Function is better than hard coded data as
1. You can easily change the behavior of data, i.e. if changing AccessKeyID from numeric -> alpha numeric, we can do it using Random Function instead of typing the whole thing. 
2. Also, if we need to change the size of key, its always better to have a random string generator which takes size of string